Add .map chapter to dynamic task concept (#1342)
astronomer/docs: learn/dynamic-tasks.md

- Map index 2: `2300` (1000+1000+300)

## Transform outputs with .map

There are use cases in which you want to transform the output of an upstream task before another task dynamically maps over it. For example, if the upstream traditional operator returns its output in a fixed format or if you want to skip certain mapped task instances based on a logical condition.

The `.map()` method was added in Airflow 2.4 and allows you to use any Python function to transform an iterable input before a task dynamically maps over it.

You can call `.map()` directly on a task using the TaskFlow API (`my_upstream_task_flow_task().map(mapping_function)`) or on the output object of a traditional operator (`my_upstream_traditional_operator.output.map(mapping_function)`).

The downstream task is dynamically mapped over the object created by the `.map()` method using either `.expand()` for a single keyword argument or `.expand_kwargs()` for list of dictionaries containing sets of keyword arguments.

The code snippet below shows how to use `.map()` to skip specific mapped tasks based on a logical condition.

- `list_strings` is the upstream task returning a list of strings.

> - `skip_strings_starting_with_skip` is a mapping function transforming the list of strings into a list of modified strings and `AirflowSkipException`s. This function will not show up as an Airflow task.

Reviewer: ```suggestion
-`skip_strings_starting_with_skip` transforms a list of strings into a list of modified strings and `AirflowSkipExceptions`. In this DAG, the function transforms `list_strings` into a new list called `transformed_list`. This function will not appear as an Airflow task.
```

> - `mapped_printing_function` dynamically maps over the `transformed_list` object created by using the `.map()` method with the mapping function (`skip_strings_starting_with_skip`) as an argument on the upstream task `list_strings`.

Reviewer: ```suggestion
- `mapped_printing_function` dynamically maps over the `transformed_list` object.
```
By explaining the previous bullet point again within this bullet, it starts to muddy the waters for what specific behavior is being described. Also, does `mapped_printing_function` appear as a task? Is that a meaningful distinction to note?

> - `mapped_printing_function` dynamically maps over the `transformed_list` object created by using the `.map()` method with the mapping function (`skip_strings_starting_with_skip`) as an argument on the upstream task `list_strings`.

Author: ```suggestion
- `mapped_printing_task` dynamically maps over the `transformed_list` object created by using the `.map()` method with the mapping function (`skip_strings_starting_with_skip`) as an argument on the upstream task `list_strings`.
```

```python
# an upstream task returns a list of outputs in a fixed format
@task
def list_strings():
    return ["skip_hello", "hi", "skip_hallo", "hola", "hey"]

# the function used to transform the upstream output before 
# a downstream task is dynamically mapped over it
def skip_strings_starting_with_skip(string):
    if len(string) < 4:
        return string + "!"
    elif string[:4] == "skip":
        raise AirflowSkipException(
            f"Skipping {string}; as I was told!"
        )
    else:
        return string + "!"
        

> # transforming the output of the first task with the map function

Reviewer: ```suggestion
# Transforming the output of the first task with the map function.
```

> # for non-TaskFlow operators use 

Reviewer: ```suggestion
# For non-TaskFlow operators, use 
```

# my_upstream_traditional_operator.output.map(mapping_function)
transformed_list = list_strings().map(skip_strings_starting_with_skip)

> # the task using dynamic task mapping on the transformed list of strings

Author: ```suggestion
# Use dynamic task mapping on the transformed list of strings.
```
